### PR TITLE
Adds schema2proto cli utility

### PR DESCRIPTION
**Pull request by cromwellian**
Usage: ./tools/schema2proto --out <destDir> <manifest files>

e.g.
Run ./tools/schema2proto --out proto particles/canonical.manifest
from Arcs repo root.

Limitations:
Currently only primitive fields.
Doesn't support overriding default package name (hardcoded 'arcs')
Should support .proto import directive to de-dup URL message and
other shared proto definitions.

---
This change is [Reviewable](https://reviewable.io/reviews/polymerlabs/arcs/3007)

### CODE REVIEW

**Comment by cromwellian** [comment]
Output will look like this: 
ls proto/
ArtistFind.proto	Description.proto	Friend.proto		Person.proto		Restaurant.proto	TVMazeShow.proto
Avatar.proto		Event.proto		Geolocation.proto	Playlist.proto		Stats.proto		Text.proto
Board.proto		FavoriteFood.proto	Move.proto		Product.proto		TVMazeQuery.proto	UserName.proto

$ cat proto/TVMazeShow.proto

syntax = "proto2";

package arcs;

message TVMazeShow {

    optional string day = 1;
    optional bool delete = 2;
    optional string description = 3;
    optional string extra = 4;
    optional bool favorite = 5;
    optional string image = 6;
    optional string name = 7;
    optional string network = 8;
    optional string showid = 9;
    optional string suggestion = 10;
    optional string time = 11;
}

File: src/runtime/wasm-tools.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Schema} from './schema.js';
+import { toProtoJSON } from './wasm.js';
+import proto_target from 'protobufjs/cli/targets/proto.js';
+import Root from 'protobufjs/src/root.js';
+
+// Convert a Schema to .proto format that can be used to compile protobuf wrappers
+export async function toProtoFile(schema: Schema) {
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply src/runtime/wasm-tools.ts]: It's a text proto description. Yes, I could assert that the text output equals a handcoded fixture, but it would mean updating the fixture everytime the test inputs are updated and wouldn't test if the proto def actually round trips. 

What looking at working on next is a C++ wasm manual integration test that if node is launched with --expose wasm, we can test the round trip from protobufjs to C++ protobuf compiler.  I'm happy to nix the roundtrip on this test if you want in favor of a C++ end-to-end test.

File: src/runtime/wasm-tools.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/**
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply src/runtime/wasm-tools.ts]: I didn't put it there because I wasn't sure if it's going to pull in a lot of unneeded dependencies into the runtime bundle. Since toProtoFile() was for cli tools not runtime and relies on src/cli/targets/... package, which is mostly needed for the command like pbjs/ts, I thought keeping it separate makes it easier to exclude and track.

**Comment by cromwellian** [comment]
For the src/root.js issue, I think it might exist on protobufjs.Root, I'll check.

File: src/runtime/wasm-tools.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mykmartin) [reply src/runtime/wasm-tools.ts]: I think we want a fairly well-defined translation from schema fields to proto fields, and a unit test that lays that out explicitly would be useful. Having to update the test to match when the logic changes is a feature, IMO.

Testing that the proto round-trips is a separate concern (and we can have that, but as a separate test).

File: src/runtime/wasm-tools.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mykmartin) [reply src/runtime/wasm-tools.ts]: Fair enough.

File: tools/schema2proto [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+DIR=$(dirname $SCRIPTPATH)
+cd $DIR
+exec ./build/cli/schema2proto.js "$@"
```
[the remark this answers is not in the document]
Comment (mykmartin) [reply tools/schema2proto]: Just had a chat with the runtime team; it's preferable to integrate tools like this into sigh for invocation. You can use the 'bundle' tool as a guide: see `src/tools/bundle-cli.ts` the `bundle` function in sigh.ts.

File: src/cli/schema2proto.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ */
+
+import fs from 'fs';
+import {Manifest} from '../runtime/manifest.js';
+import {Schema} from '../runtime/schema.js';
+import {toProtoFile} from '../runtime/wasm-tools.js';
+import {Utils} from '../../shells/lib/runtime/utils.js';
+
+// Converts schema definitions specifies in Manifest files to proto2 specifications.
+void (async () => {
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply src/cli/schema2proto.ts]: I copied the template from spotify-importer.ts, I assumed it was to suppress some kind of lint check, but it works without it, so removed.

File: tools/schema2proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (cromwellian) [reply tools/schema2proto]: Thanks, I didn't know about this tool. Cool.

File: src/cli/schema2proto.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env -S node --experimental-modules --no-deprecation --loader=./tools/custom-loader.mjs -r source-map-support/register.js
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply src/cli/schema2proto.ts]: Removed in favor of tools/sigh launcher

**Comment by cromwellian** [comment]
Now runnable as tools/sigh schema2proto -o outputDir particles/canonical.manifest

File: particles/Native/Java/source/j2clparticle.iws [hunk 6]
```diff
@@ -2,8 +2,7 @@
 <project version="4">
```
Comment (shaper) [inline particles/Native/Java/source/j2clparticle.iws]: Do we need the `.iws` file in the repo?

File: src/runtime/wasm.ts [hunk 7]
```diff
@@ -44,7 +44,7 @@ function jsonBaseType(type) {
 }
 
 // Returns a JSON representation that protobufjs can use to de/serialize entity data as protobufs.
-function toProtoJSON(schema: Schema) {
+export function toProtoJSON(schema: Schema) {
```
Comment (shaper) [inline src/runtime/wasm.ts]: Is @mykmartin good with this? It looks like `wasm.ts` aims to only export `EntityProtoConverter`, maybe he'd prefer this to be exported as a method on that class? Can do as needed in separate PR since he's already approved.

Comment (mykmartin) [reply src/runtime/wasm.ts]: This is fine. I expect the shape of this code to change a lot as the wasm implementation progresses, so I'm not worried about specifics yet.
